keps/sig-node/20200303-shared-pod-limits.md

```diff
@@ -0,0 +1,325 @@
+
+### Goals
+
+This proposal aims to:
+
+* Allow developers to define resource limits on the pod level in addition to the individual container level.
+* Use as much as possible the Linux kernel's cgroup ability to define limits in hierarchies to enforce the defined limits.
+
+### Non-Goals
+
+*  Providing a general-purpose interface to the full range of possible resource management provided by the Linux cgroup hierarchy.
+
+## Proposal
+
+The Pod QoS enhancement already implemented in Kubernetes manages resources as a hieararchy of cgroups in the following way:
+
+<pre>
+  kubepods
+   |
+   +-- Guaranteed-pod0
+   |   |
+   |   +-- container0 (pause container)
+   |   |
+   |   +-- container1 (first container)
+   |   |
+   |  = ... =
+   |   |
+   |   +-- containerN (N-th container)
+   |
+   +-- QoS CGroup (one of burstable, or besteffort)
+       |
+       \ pod 
+          |
+          +-- container0 (pause container)
+          |
+          +-- container1 (first container)
+          |
+        = ... =
+          |
+          +-- containerN (N-th container)
+</pre>
+
+The current implementation will set the cgroup limits for the memory/CPU resources on the pod level of the hierarchy only in the following cases:
+1. The QoS level is `Guaranteed`. 
+1. The QoS level is `Burstable` and all of the containers specify a limit for the relevant resource. 
+
+For the current definition of the `Guaranteed` QoS level this proposal would not make any modifications. Since each container is provided with its
+own limit, defining a pod-level limit is redundant. If the pod level limit request is smaller than the sum of limit for all the containers, then 
+the limit requested by the containers is unreachable. Conversely if the pod level limit request is larger than the sum of the limit for all the 
+containers then the excess is not relevant since the individual containers can never use more than their requested share, and even if all of the 
+containers use the maximum allowed by their defined limit, it can never amount to the pod-level limit.
+
+For pods in the `Burstable` and `BestEffort` QoS levels specifying a resource limit on the pod level means that resources can be shared across containers.
+Sharing resources across containers is possible only if individual containers release the resources that are not being used. For CPU this is trivial.
+For memory, this depends on the application being run actually releasing the memory back to the operating system. See [below](#Memory-Reuse-Across-cgroups)
+for an analysis.
+
+The proposal in this KEP is to allow users to define resources on the Pod level itself. Only limits will be allowed on this level so that the effect on
```

Comment:
If only limits are allowed, then I can't use this to allocate a "slush fund" of resources that I can share between my containers, which is a pretty compelling use-case in my mind..

Reply:
Unfortunately the Linux cgroup controllers don't support this sort of behavior at all for memory. The `request` section of a `ResourceList` in a container can't be used for setting up memory reservations even on the container level. It's simply used by the Kubernetes scheduler as a parameter for placing the pod. If the pod doesn't allocate the memory as soon as it's running, there's no guarantee that it will be able to allocate the memory in the future. It's not even guaranteed to succeed if the pod tries to do this at once on startup. The decision the scheduler makes when placing the pod is only correct for the point in time when the scheduling algorithm ran - the memory situation on the node could have changed by the time the pod actually started running. 

The most you can do with the memory cgroup controller is set a `memory.soft_limit_in_bytes` which is used as a hint for the kernel's memory subsystem around how much memory to try (as in best-effort) to leave allocated to the cgroup once an OOM event occurs. See the memory cgroup controller [documentation](https://www.kernel.org/doc/Documentation/cgroup-v1/memory.txt) for the nitty-gritty details. There's no technical way to setup a "slush-fund" in the way you describe for memory.

The behavior you outlined for the CPU cgroup is what you get by setting the kubernetes limit for the pod, and using the request on the individual container level for setting up the shares allocated to each leaf cgroup. It's exactly this, and it's exactly the behavior you would get with this KEP. The difference between before and after this KEP is that before this KEP, the pod would not have a "slush fund" limit, so all containers in the QoS group basically compete with one another. With this KEP, the containers of a pod compete with each other first, and then with other pods in the same QoS level.

Reply:
On memory.  It is possible, with caveats, to get some utility here.

Consider a pod with 2 containers:
  1: {requests: 100 MiB, limits: 200 MiB}
  2: {requests: 200 MiB, limits: 300 MiB}

Scheduler today would allow this to run on a node with just 300 MiB available (perfect fit).

If I set the pod resources as {requests: 400 MiB, limits: 400 MiB}, my pod is still guaranteed class, though my containers might OOM.  I believe the kernel will even do the righ tthing on reclaim here (try to squeeze within the parent cgroup), but it's been a while since I was deep in that subsystem, so I am not 100% sure.  The main point is that the pod will no longer schedule into 300 MiB, because we requested 400.  The containers can use that "extra" 100 MiB as the slush fund.  Spike into it when needed.

CPU is behaviorally different, but semantically the same - if I want to allocate some CPU that is guaranteed available to my pod, I need to specify it at the pod level.

I think it is semantically consistent (though I know there are odd cases in the kernel subsystems).

Reply:
What you described is the exact behavior you would get with container requests of 200Mi and 200Mi today (assuming we introduced pod-level limits).  The kubelet sets both oom_score_adj and performs eviction based on pod-level memory requests, so the container-level ones don't end up mattering.  I.E. 100 + 300 is equivalent to 200 + 200.

Reply:
That is defining the API in terms of the implementation.  It loses the expression of intent.  If we successfully port some of our internal kernel patches upstream, the request might actually matter, and our API won't allow users to express what they meant.

Eg we might set soft-limit to request, and then it actually DOES kind of matter - the reclaim logic is influenced by that.

Reply:
What would be the advantage of adding a `requests` section to the pod level right now, when there's nothing in the implementation that would benefit from it? It would mean that the scheduler would also need to change to account for the new `requests` section, but it wouldn't add anything that isn't already achievable without it.

If any kernel patches make it upstream and it makes sense to add this in the future - then it can be added in the future.

Although I think setting the memory soft-limit on the container level based on the request value in the container level makes sense already today, once pod level limits are defined.

Reply:
Look again at the example:


Pod with 2 containers:
c1: {requests: 100 MiB, limits: 200 MiB}
c2: {requests: 200 MiB, limits: 300 MiB}
Pod: {requests: 400 MiB, limits: 400 MiB}

The pod is effectively guaranteed class.  Semantically the intention is clear: 100 for c1, 200 for c2, and 100 to be shared between them.  I *could* make them 200+200 but that forces the user to "lie" (it's not what they really wanted to express) for the sake of the implementation.

And yes, the scheduler would have to consider pod requests, though I think we need to think about whether that field would be defaulted or left empty with well-defined meaning.  Especially in light of the in-place-update KEP.

Reply:
I really agree with @thockin that adding `requests` as well as `limits` is a quite compelling use case.

It would essentially make it possible to say that if a podspec has `requests == limits` the pod will become a Guaranteed pod as in qos (if we choose to do that), even tho the containers don't have that strict requirements. That will make it easier for pods with sidecars to avoid wasting unused reserved resources while still being a guaranteed pod.

The same kinda applies to burstable and besteffort pods as well. When a besteffort pod gets a pod-level limit, then it is essentially a burstable pod. Should we then still treat it as a besteffort pod?

Reply:
I'm trying to wrap my head around the semantics that pod-level resource limits will impose on the `CPUManager` and `DeviceManager`.

At present with the `CPUManager` `static` policy enabled, it operates as follows:
1. By default, all non-Guaranteed containers are placed into a „bounding“ cset cgroup containing all CPUs on the system (minus any „reserved“ CPUs set up at kubelet launch)
2. At pod admission time, the `CPUManager` checks to see if a pod is in the Guaranteed QoS class (i.e. all containers have their respective cpu and memory requests == limits)
3. If it is, then for each container requesting a whole number of CPUs (i.e 1, 2, 1000m, etc.) a. Allocate a subset set of CPUs to that container (from the bounding set) and isolate them into their own cset b. Update the „bounding“ cset for all non-Guaranteed containers to remove the newly allocated CPUs
4. For each container *not* requesting a whole number of CPUs, treat them as non-Guaranteed and associate them with the „bounding“ cset
5. Whenever a Guaranteed container finishes, free the CPUs previously granted to it, and update the „bounding“ cset for all non-Guaranteed containers to include these CPUs again

With pod-level resource limits, we will need to work out the details for how these limits interact with the semantics above.
At the moment I don’t seen an obvious answer, unfortunately.

There are a couple of dimensions to consider:
1. What constitutes a „Guaranteed“ pod with a mixture of pod-limits and container-limits
2. What happens when pod limits are set, and no container limits are set
3. What happens when pod limits are set, and *some* container limits are set to whole number and others aren’t
4. What happens when pod limits are set, and the sum of the container limits don’t equal the pod limits

As a user of the `CPUManager` I think the most intuitive use of pod-level limits would be to follow a *one-or-the-other* approach.

Meaning you either:
1. Have pod-level CPU limits set; or
2. You have container level CPU limits set;
3. But not both

Having *only’ container-level CPU limits would continue to operate as above.
And having pod-level CPU limits would operate similarly, except that *all* containers in the pod would share access to the same exclusive cset instead of them each getting their own.

These semantics then extend nicely to the `DeviceManager` where setting a pod-level limit implies that all containers share access to a device, and container-level limits implies a container gets exclusive access.
I’m not actually sure how it would work with devices otherwise.

Reply:
As far as the `DeviceManager` goes - I would consider this out-of-scope for this KEP, and I would keep the current behavior.  As specified in the [documentation](https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/#device-plugin-registration), devices cannot be shared between containers. Does it make sense for two containers in the same pod to try to allocate the same GPU?

When considering the `CPUManager`: If a pod is guaranteed because it has a pod-level request that matches the pod-level limit, then the `CPUManager` can ignore the containers in the pod and perform the same exact algorithm it would have done for all of the containers - except exclusively on the pod level cgroup. The effect would be the same consistent with the explicit purpose of the `CPUManager`; all of the containers in that cgroup sub-hierarchy would be isolated to the same CPU cset. The limits on the container(s) level in the pod would be used for managing the `cpuacct` cgroup only.

Reply:
Regarding:
> Does it make sense for two containers in the same pod to try to allocate the same GPU?
Yes. This is a common request, and we've wanted something like pod-level resource limits to enable this for quite some time. At present we hack around this by letting people set environment variables to enable the sharing (thus bypassing the Kubernetes resource model completely), but it would be much cleaner if they could leverage pod-level resource limits to do this instead.

Regarding the `CPUManager` semantics you propose -- I think this makes alot of sense and is the way we should implement it. So long as `request == limit` at the pod-level, all containers in the pod share access to the same set of exclusive CPUs (with the container-level requests/limits only enforced as cpushares and not csets).

Do we know yet what the formal definition of inclusion in the `Guaranteed` QoS class will be with pod-level resources?
Right now it's set when *all* containers have `requests == limits` (even init containers).

Reply:
I'm not sure this specific KEP is a good match for solving the `DeviceManager` problem. Is the intent really to share a device across _all_ containers in a pod, or about sharing a device between a specific subset of containers in the pod? If something like Istio injects a sidecar into a pod, should a GPU device also be shared with the Istio sidecar - just because the device is mentioned on the pod level? 

Wouldn't device sharing be better handled by allowing the developer to specify somehow exactly which subset of containers should have access to which device by marking the containers in the sharing set, rather than by moving the device share up the pod hierarchy?

> Do we know yet what the formal definition of inclusion in the Guaranteed QoS class will be with pod-level resources?
Right now it's set when all containers have requests == limits (even init containers).

A pod would be in the Guaranteed QoS if either:
* The pod has `requests == limits` for both memory and cpu. There is no requirement on the containers in the pod in this case.

or

* The pod doesn't define any pod-level resources and all of the pod containers (including init containers) have have `requests == limits` for both memory and cpu (like before this kep)

Reply:
> I'm not sure this specific KEP is a good match for solving the DeviceManager problem

That is fair. So long as we are explicit about this. So pod-level limits will only be allowed for `cpu` and `memory`?
What will happen if someone does set a limit on a device resource (or any other custom resource for that matter)?

> A pod would be in the Guaranteed QoS if either ...

That seems reasonable -- and it matches what we said we would do in the `CPUManager` when / if pod-level limits were set for CPUs.